### ISSUE TITLE
Prefer Guard Clauses to Nesting

### ISSUE THREAD

**Issue by nbibler** [HaulinOats/todo-app#6]
We have a general preference to use guard clauses rather than nesting significant logic inside of conditionals. This tends to make code easier to read and reasonable about.

For example:

https://github.com/HaulinOats/todo-app/blob/9eb0ce07afc81361417f443b4872c7c6a3b47016/components/TodoList.tsx#L38-L49

Could alternatively be written as:

```typescript
  const addTodo: KeyboardEventHandler<HTMLInputElement> = (event) => {
    const target = event.currentTarget;

    // There's a bug here, the `target.value` should likely be `trim()`'d to protect against whitespace.
    if (event.key !== 'Enter' || target.value.length === 0) {
      return;
    }

    setTodos([...todos, {
      id: Date.now(),
      label: target.value,
      isCompleted: false
    }])

    target.value = '';
  }
```

Also note the use of [`currentTarget`](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget) there, rather than `target`. `currentTarget` is guaranteed to be the originating element (HTMLInputElement) in this case. Whereas `target` could be an element that's been bubbled up to (and why TypeScript couldn't tell you what it was).

You can also simplify the reference in the template a bit:

```diff
-  onKeyDown={e => addTodo(e)} />
+  onKeyDown={addTodo} />
```

**Comment by HaulinOats** [comment]
I'm still getting the hang of Typescript but all good to know. I'm still trying to understand the differences between all the different event typings but that particular syntax makes it a lot easier.

I also struggle a little still with the syntax on declaring event handlers on elements because I first learned react before hooks were a thing so you could either bind functions to the class itself (which felt needlessly verbose having to bind every function in the constructor), or you could write inline arrow functions. 

So basically, if there's only the event itself, you can just define it like an IIFE `onKeyDown={addTodo} />`, otherwise, you would need to defer to arrow function syntax (to allow for additional parameters)? This is one of those things where I should understand better what's going on under the hood. Again, still have a little confusion over how functional components handle event handlers versus the class-based way.
